payjoin-cli/src/app.rs

```diff
@@ -318,23 +319,23 @@ impl App {
             .get_new_address(None, None)
             .map_err(|e| Error::Server(e.into()))?
             .assume_checked();
-        payjoin.substitute_output_address(receiver_substitute_address);
-
-        let payjoin_proposal_psbt = payjoin.apply_fee(Some(1))?;
+        provisional_payjoin.substitute_output_address(receiver_substitute_address);
 
-        log::debug!("Extracted PSBT: {:#?}", payjoin_proposal_psbt);
-        // Sign payjoin psbt
-        let payjoin_base64_string = base64::encode(&payjoin_proposal_psbt.serialize());
-        // `wallet_process_psbt` adds available utxo data and finalizes
-        let payjoin_proposal_psbt = self
-            .bitcoind
-            .wallet_process_psbt(&payjoin_base64_string, None, None, Some(false))
-            .map_err(|e| Error::Server(e.into()))?
-            .psbt;
-        let payjoin_proposal_psbt = Psbt::from_str(&payjoin_proposal_psbt)
-            .context("Failed to parse PSBT")
-            .map_err(|e| Error::Server(e.into()))?;
-        let payjoin_proposal_psbt = payjoin.prepare_psbt(payjoin_proposal_psbt)?;
+        let payjoi_proposal = provisional_payjoin.finalize_proposal(
+            |psbt: &Psbt| {
+                self.bitcoind
+                    .wallet_process_psbt(
+                        &bitcoin::base64::encode(psbt.serialize()),
+                        None,
+                        None,
+                        Some(false),
+                    )
+                    .map(|res| Psbt::from_str(&res.psbt).unwrap())
```

Comment:
@DanGould been trying to tackle this `unwrap` without success. how would you approach it?

Reply:
This article gives a few different options for returning errors from iterators: https://doc.rust-lang.org/rust-by-example/error/iter_result.html

I think the main issue is that it it doesn't make sense for `finalize_proposal` to return a `RequestError`, because the request has already been checked by this point. The error is more of a receiver::Error::Server and could be returned inside that Box.

Reply:
ah changing that to Error made things so easier! thanks